Add minimal x86_64-lynx-lynxos178 support.

tnewsome-lynx opened the pull request:
Add minimal x86_64-lynx-lynxos178 support. It's possible to build no_std
programs with this compiler.

## Tier 3 Target Policy

> A tier 3 target must have a designated developer or developers (the "target
maintainers") on record to be CCed when issues arise regarding the target. (The
mechanism to track and CC such developers may evolve over time.)

Tim Newsome (@tnewsome-lynx) will be the designated developer for
x86_64-lynx-lynxos178 support.

> Targets must use naming consistent with any existing targets; for instance, a
target for the same CPU or OS as an existing Rust target should use the same
name for that CPU or OS. Targets should normally use the same names and naming
conventions as used elsewhere in the broader ecosystem beyond Rust (such as in
other toolchains), unless they have a very good reason to diverge. Changing the
name of a target can be highly disruptive, especially once the target reaches a
higher tier, so getting the name right is important even for a tier 3 target.

I believe the target is named appropriately.

> Target names should not introduce undue confusion or ambiguity unless
absolutely necessary to maintain ecosystem compatibility. For example, if the
name of the target makes people extremely likely to form incorrect beliefs about
what it targets, the name should be changed or augmented to disambiguate it.

The target name is not confusing.

> If possible, use only letters, numbers, dashes and underscores for the name.
Periods (.) are known to cause issues in Cargo.

Done.

> Tier 3 targets may have unusual requirements to build or use, but must not
create legal issues or impose onerous legal terms for the Rust project or for
Rust developers or users.
> The target must not introduce license incompatibilities.
> Anything added to the Rust repository must be under the standard Rust license
(MIT OR Apache-2.0).

All this new code is licensed under the Apache-2.0 license.

> The target must not cause the Rust tools or libraries built for any other host
(even when supporting cross-compilation to the target) to depend on any new
dependency less permissive than the Rust licensing policy. This applies whether
the dependency is a Rust crate that would require adding new license exceptions
(as specified by the tidy tool in the rust-lang/rust repository), or whether the
dependency is a native library or binary. In other words, the introduction of
the target must not cause a user installing or running a version of Rust or the
Rust tools to be subject to any new license requirements.

Done.

> Compiling, linking, and emitting functional binaries, libraries, or other code
for the target (whether hosted on the target itself or cross-compiling from
another target) must not depend on proprietary (non-FOSS) libraries. Host tools
built for the target itself may depend on the ordinary runtime libraries
supplied by the platform and commonly used by other applications built for the
target, but those libraries must not be required for code generation for the
target; cross-compilation to the target must not require such libraries at all.
For instance, rustc built for the target may depend on a common proprietary C
runtime library or console output library, but must not depend on a proprietary
code generation library or code optimization library. Rust's license permits
such combinations, but the Rust project has no interest in maintaining such
combinations within the scope of Rust itself, even at tier 3.

I think we're in the clear here. We do link against some static libraries that
are proprietary (like libm and libc), but those are not used to generate code.
E.g. the VxWorks target requires `wr-c++` to be installed, which is not
publically available.

> "onerous" here is an intentionally subjective term. At a minimum, "onerous"
legal/licensing terms include but are not limited to: non-disclosure
requirements, non-compete requirements, contributor license agreements (CLAs) or
equivalent, "non-commercial"/"research-only"/etc terms, requirements conditional
on the employer or employment of any particular Rust developers, revocable
terms, any requirements that create liability for the Rust project or its
developers or users, or any requirements that adversely affect the livelihood or
prospects of the Rust project or its developers or users.

Our intention is to allow anyone with access to LynxOS CDK to use Rust for it.

> Neither this policy nor any decisions made regarding targets shall create any
binding agreement or estoppel by any party. If any member of an approving Rust
team serves as one of the maintainers of a target, or has any legal or
employment requirement (explicit or implicit) that might affect their decisions
regarding a target, they must recuse themselves from any approval decisions
regarding the target's tier status, though they may otherwise participate in
discussions.
> This requirement does not prevent part or all of this policy from being cited
in an explicit contract or work agreement (e.g. to implement or maintain support
for a target). This requirement exists to ensure that a developer or team
responsible for reviewing and approving a target does not face any legal threats
or obligations that would prevent them from freely exercising their judgment in
such approval, even if such judgment involves subjective matters or goes beyond
the letter of these requirements.

No problem.

> Tier 3 targets should attempt to implement as much of the standard libraries
as possible and appropriate (core for most targets, alloc for targets that can
support dynamic memory allocation, std for targets with an operating system or
equivalent layer of system-provided functionality), but may leave some code
unimplemented (either unavailable or stubbed out as appropriate), whether
because the target makes it impossible to implement or challenging to implement.
The authors of pull requests are not obligated to avoid calling any portions of
the standard library on the basis of a tier 3 target not implementing those
portions.

With this first PR, only core is supported. I am working on support for the std
library and intend to submit that once all the tests are passing.

> The target must provide documentation for the Rust community explaining how to
build for the target, using cross-compilation if possible. If the target
supports running binaries, or running tests (even if they do not pass), the
documentation must explain how to run such binaries or tests for the target,
using emulation if possible or dedicated hardware if necessary.

This is documented in `src/doc/rustc/src/platform-support/lynxos178.md`.

> Tier 3 targets must not impose burden on the authors of pull requests, or
other developers in the community, to maintain the target. In particular, do not
post comments (automated or manual) on a PR that derail or suggest a block on
the PR based on a tier 3 target. Do not send automated messages or notifications
(via any medium, including via @) to a PR author or others involved with a PR
regarding a tier 3 target, unless they have opted into such messages.
> Backlinks such as those generated by the issue/PR tracker when linking to an
issue or PR are not considered a violation of this policy, within reason.
However, such messages (even on a separate repository) must not generate
notifications to anyone involved with a PR who has not requested such
notifications.

Understood.

> Patches adding or updating tier 3 targets must not break any existing tier 2
or tier 1 target, and must not knowingly break another tier 3 target without
approval of either the compiler team or the maintainers of the other tier 3
target.
> In particular, this may come up when working on closely related targets, such
as variations of the same architecture with different features. Avoid
introducing unconditional uses of features that another variation of the target
may not have; use conditional compilation or runtime detection, as appropriate,
to let each target run code supported by that target.

As far as I know this change does not affect any other targets.

> Tier 3 targets must be able to produce assembly using at least one of rustc's
supported backends from any host target. (Having support in a fork of the
backend is not sufficient, it must be upstream.)

Many targets produce assembly for x86_64 so that also works for LynxOS-178.

tnewsome-lynx committed:
Add minimal x86_64-lynx-lynxos178 support.

It's possible to build no_std programs with this compiler.

> A tier 3 target must have a designated developer or developers (the "target
maintainers") on record to be CCed when issues arise regarding the target. (The
mechanism to track and CC such developers may evolve over time.)

Tim Newsome (@tnewsome-lynx) will be the designated developer for
x86_64-lynx-lynxos178 support.

> Targets must use naming consistent with any existing targets; for instance, a
target for the same CPU or OS as an existing Rust target should use the same
name for that CPU or OS. Targets should normally use the same names and naming
conventions as used elsewhere in the broader ecosystem beyond Rust (such as in
other toolchains), unless they have a very good reason to diverge. Changing the
name of a target can be highly disruptive, especially once the target reaches a
higher tier, so getting the name right is important even for a tier 3 target.

I believe the target is named appropriately.

> Target names should not introduce undue confusion or ambiguity unless
absolutely necessary to maintain ecosystem compatibility. For example, if the
name of the target makes people extremely likely to form incorrect beliefs about
what it targets, the name should be changed or augmented to disambiguate it.

The target name is not confusing.

> If possible, use only letters, numbers, dashes and underscores for the name.
Periods (.) are known to cause issues in Cargo.

Done.

> Tier 3 targets may have unusual requirements to build or use, but must not
create legal issues or impose onerous legal terms for the Rust project or for
Rust developers or users.
> The target must not introduce license incompatibilities.
> Anything added to the Rust repository must be under the standard Rust license
(MIT OR Apache-2.0).

All this new code is licensed under the Apache-2.0 license.

> The target must not cause the Rust tools or libraries built for any other host
(even when supporting cross-compilation to the target) to depend on any new
dependency less permissive than the Rust licensing policy. This applies whether
the dependency is a Rust crate that would require adding new license exceptions
(as specified by the tidy tool in the rust-lang/rust repository), or whether the
dependency is a native library or binary. In other words, the introduction of
the target must not cause a user installing or running a version of Rust or the
Rust tools to be subject to any new license requirements.

Done.

> Compiling, linking, and emitting functional binaries, libraries, or other code
for the target (whether hosted on the target itself or cross-compiling from
another target) must not depend on proprietary (non-FOSS) libraries. Host tools
built for the target itself may depend on the ordinary runtime libraries
supplied by the platform and commonly used by other applications built for the
target, but those libraries must not be required for code generation for the
target; cross-compilation to the target must not require such libraries at all.
For instance, rustc built for the target may depend on a common proprietary C
runtime library or console output library, but must not depend on a proprietary
code generation library or code optimization library. Rust's license permits
such combinations, but the Rust project has no interest in maintaining such
combinations within the scope of Rust itself, even at tier 3.

I think we're in the clear here. We do link against some static libraries that
are proprietary (like libm and libc), but those are not used to generate code.
E.g. the VxWorks target requires `wr-c++` to be installed, which is not
publically available.

> "onerous" here is an intentionally subjective term. At a minimum, "onerous"
legal/licensing terms include but are not limited to: non-disclosure
requirements, non-compete requirements, contributor license agreements (CLAs) or
equivalent, "non-commercial"/"research-only"/etc terms, requirements conditional
on the employer or employment of any particular Rust developers, revocable
terms, any requirements that create liability for the Rust project or its
developers or users, or any requirements that adversely affect the livelihood or
prospects of the Rust project or its developers or users.

Our intention is to allow anyone with access to LynxOS CDK to use Rust for it.

> Neither this policy nor any decisions made regarding targets shall create any
binding agreement or estoppel by any party. If any member of an approving Rust
team serves as one of the maintainers of a target, or has any legal or
employment requirement (explicit or implicit) that might affect their decisions
regarding a target, they must recuse themselves from any approval decisions
regarding the target's tier status, though they may otherwise participate in
discussions.
> This requirement does not prevent part or all of this policy from being cited
in an explicit contract or work agreement (e.g. to implement or maintain support
for a target). This requirement exists to ensure that a developer or team
responsible for reviewing and approving a target does not face any legal threats
or obligations that would prevent them from freely exercising their judgment in
such approval, even if such judgment involves subjective matters or goes beyond
the letter of these requirements.

No problem.

> Tier 3 targets should attempt to implement as much of the standard libraries
as possible and appropriate (core for most targets, alloc for targets that can
support dynamic memory allocation, std for targets with an operating system or
equivalent layer of system-provided functionality), but may leave some code
unimplemented (either unavailable or stubbed out as appropriate), whether
because the target makes it impossible to implement or challenging to implement.
The authors of pull requests are not obligated to avoid calling any portions of
the standard library on the basis of a tier 3 target not implementing those
portions.

With this first PR, only core is supported. I am working on support for the std
library and intend to submit that once all the tests are passing.

> The target must provide documentation for the Rust community explaining how to
build for the target, using cross-compilation if possible. If the target
supports running binaries, or running tests (even if they do not pass), the
documentation must explain how to run such binaries or tests for the target,
using emulation if possible or dedicated hardware if necessary.

This is documented in `src/doc/rustc/src/platform-support/lynxos_178.md`.

> Tier 3 targets must not impose burden on the authors of pull requests, or
other developers in the community, to maintain the target. In particular, do not
post comments (automated or manual) on a PR that derail or suggest a block on
the PR based on a tier 3 target. Do not send automated messages or notifications
(via any medium, including via @) to a PR author or others involved with a PR
regarding a tier 3 target, unless they have opted into such messages.
> Backlinks such as those generated by the issue/PR tracker when linking to an
issue or PR are not considered a violation of this policy, within reason.
However, such messages (even on a separate repository) must not generate
notifications to anyone involved with a PR who has not requested such
notifications.

Understood.

> Patches adding or updating tier 3 targets must not break any existing tier 2
or tier 1 target, and must not knowingly break another tier 3 target without
approval of either the compiler team or the maintainers of the other tier 3
target.
> In particular, this may come up when working on closely related targets, such
as variations of the same architecture with different features. Avoid
introducing unconditional uses of features that another variation of the target
may not have; use conditional compilation or runtime detection, as appropriate,
to let each target run code supported by that target.

As far as I know this change does not affect any other targets.

> Tier 3 targets must be able to produce assembly using at least one of rustc's
supported backends from any host target. (Having support in a fork of the
backend is not sufficient, it must be upstream.)

Many targets produce assembly for x86_64 so that also works for LynxOS-178.

Noratrieb commented on compiler/rustc_target/src/spec/targets/x86_64_unknown_lynxos_178_elf.rs:
```diff
@@ -0,0 +1,36 @@
+// Starting with x86_64_unknown_linux_gnu
```
this looks like a mistake

tnewsome-lynx replied:
Thanks. Removing.

Noratrieb commented on compiler/rustc_target/src/spec/targets/x86_64_unknown_lynxos_178_elf.rs:
```diff
@@ -0,0 +1,36 @@
+        llvm_target: "x86_64-unknown-unknown-gnu".into(),
+        metadata: crate::spec::TargetMetadata {
+            description: Some("LynxOS-178".into()),
+            tier: Some(1),
```
3

tnewsome-lynx replied:
Thank you. Done.

Noratrieb commented on compiler/rustc_target/src/spec/targets/x86_64_unknown_lynxos_178_elf.rs:
```diff
@@ -0,0 +1,36 @@
+        metadata: crate::spec::TargetMetadata {
+            description: Some("LynxOS-178".into()),
+            tier: Some(1),
+            host_tools: Some(true),
```
`Some(false)`

tnewsome-lynx replied:
Thank you. Also done.

Noratrieb commented on compiler/rustc_target/src/spec/targets/x86_64_unknown_lynxos_178_elf.rs:
```diff
@@ -0,0 +1,36 @@
+            description: Some("LynxOS-178".into()),
+            tier: Some(1),
+            host_tools: Some(true),
+            std: Some(true),
```
not yet true, but since you aim for this to be eventually true this seems fine

tnewsome-lynx replied:
Now that you've pointed it out, I prefer to make this false. It's not true with this commit. When it becomes true (in a commit involving a lot of std library) then I can flip it to true.

Noratrieb commented on src/doc/rustc/src/platform-support/lynxos_178.md:
```diff
@@ -0,0 +1,53 @@
+
+**Tier: 3**
+
+Targets for the LynxOS-178 operating system.
```
Can you add a short paragraph explaining what this operating system is (including a link)?

tnewsome-lynx replied:
Done.

Noratrieb commented on compiler/rustc_target/src/spec/base/lynxos_178.rs:
```diff
@@ -0,0 +1,152 @@
+    LynxosPath { env_var: &ENV_PREFIX_VAR, rel_path: "lib/crtn.o" },
+];
+
+pub(crate) fn check_environment() -> Result<(), String> {
```
We don't want such eager checking. There's no reason why `rustc --target $target --emit obj` would need these checks.

I'd prefer to not have any advance checks and just let linking fail if something is wrong. While checks are friendlier for the user, they make the code more complicated and are, as mentioned above, prone to firing unnecessarily.

Noratrieb commented on compiler/rustc_target/src/spec/base/lynxos_178.rs:
```diff
@@ -0,0 +1,152 @@
+pub(crate) fn opts() -> TargetOptions {
+    let linker_flavor = LinkerFlavor::Gnu(Cc::No, Lld::No);
+    let other_linker_flavor = LinkerFlavor::Gnu(Cc::No, Lld::Yes);
+    let env_prefix = env::var("ENV_PREFIX").unwrap_or("ENV_PREFIX".into());
```
Requiring `ENV_PREFIX` to be set won't work when just loading targets without the intention of compiling got LynxOS. You either need a dummy fallback or preferably not have it at all.
How is your toolchain expected to be used that introduces these requirements? I assume you have the same env var for C code already today.

Noratrieb reviewed (changes requested):
For the target name, `x86_64-unknown-lynxos_178-elf` is fine, but is there any reason to have the elf in the end? It seems unnecessary, unless LynxOS also supports other binaries (we could also write lynx in the vendor field but thats not very important).

tnewsome-lynx commented:
I'll be looking more into the CI failure, but locally it's passing:
```
> ./x test --target x86_64-unknown-linux-gnu assembly
...
test result: ok. 499 passed; 0 failed; 36 ignored; 0 measured; 0 filtered out; finished in 3.74s
```

tnewsome-lynx commented:
@Noratrieb Thanks for the review. Do you prefer I keep the history clean and force push, or do you prefer I make new git commits for your feedback?

Noratrieb commented:
We want to squash it together before merging, but before that it's up to you. Force pushing is fine.

jieyouxu commented:
While still ironing out the review process... https://forge.rust-lang.org/compiler/proposals-and-stabilization.html#targets now says adding new target needs r+ from compiler-leads.

So I guess cc @davidtwco and @wesleywiser for this proposal.

bors commented:
:umbrella: The latest upstream changes (presumably #135651) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

jieyouxu commented:
r? compiler_leads

bjorn3 commented on compiler/rustc_target/src/spec/base/lynxos_178.rs:
```diff
@@ -0,0 +1,80 @@
+        crt_static_respected: true,
+        panic_strategy: PanicStrategy::Abort,
+        // Don't rely on the path to find the correct ld.
+        linker: Some(cow_format!("{cdk_path}/bin/ld")),
```
Should this perhaps use gcc as linker driver instead? That should allow avoiding all (or at least most) pre and post link args. And instead of reading `ENV_PREFIX` and `GCC_DIR` you can have the user pass `-Clinker=/path/to/gcc` like we normally require for cross-compiling (and if LynxOS supports native compilation too, you can use the location where LynxOS puts the linker as default here).

tnewsome-lynx replied:
I'm not familiar enough with rustc to know what you mean by using "gcc as linker driver." I guess that's different from the linker_flavor TargetOption?

Use of `ENV_PREFIX` and `GCC_DIR` is baked pretty far into the LynxOS-178 development workflow. There's no way ld can find the libraries in `ENV_PREFIX` without it being set, and the same script that sets `ENV_PREFIX` also sets `GCC_DIR`.

Noratrieb replied:
The way we generally do linkage on unix-like platforms is by invoking `cc` with our libraries and object files. This will ensure that all the C standard libraries are found and we don't have to worry about them ourselves.

I assume that using `cc` (or some other GCC binary name) on LynxOS-178 still requires setting these environment variables? Or will that magically pick everything up?

tnewsome-lynx replied:
I see... I'll experiment with calling gcc instead of ld. It does seem to magically pick up the right thing, and I'm not sure what that magic depends on.

tnewsome-lynx replied:
Changing to using gcc makes the arguments much simpler, as @bjorn3 mentioned. I *think* I can get away without relying on ENV_PREFIX at all, because the LynxOS-178 setup also changes the PATH. This is good because now we can find the LynxOS-178 linker, and bad because it masks the linker for the host. I think I know how to fix this, but it's going to take a bit more talking to people and experimenting.

But I am still working on this PR, and hopefully will have something simpler for you all next week.

bjorn3 commented on compiler/rustc_target/src/spec/base/lynxos_178.rs:
```diff
@@ -0,0 +1,80 @@
+macro_rules! cow_format {
+    ($arg:expr) => {
+        // Cow::from works with &str but not with &mut str.
+        Cow::from(immutable(Box::leak(format!($arg).into_boxed_str())))
```
```suggestion
        Cow::Borrowed(Box::leak(format!($arg).into_boxed_str()))
```

tnewsome-lynx replied:
Done. Thanks.

bjorn3 commented:
The PR title says x86_64-lynxos_178-unknown-elf, but the source code says x86_64-unknown-lynxos_178-elf. I presume the latter is correct?

tnewsome-lynx commented:
> The PR title says x86_64-lynxos_178-unknown-elf, but the source code says x86_64-unknown-lynxos_178-elf. I presume the latter is correct?

Yes. Fixing that. Good eye.

davidtwco reviewed (approved):
r=me with a rebase

tnewsome-lynx commented:
There are a few changes over the last couple of weeks. The main one is that I've changed what Lynx tools are available to rustc, so the rustc code is not a lot simpler than it used to be. As part of this, I also replaced "lynxos_178" with "lynxos178" everywhere to match the triple we use when building gcc for LynxOS-178.

tnewsome-lynx commented:
@Noratrieb Can you take a look at this again? I think it's good to go now.

bors commented:
:umbrella: The latest upstream changes (presumably #139279) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rfatykhov-lynx commented:
> ☔ The latest upstream changes (presumably #139279) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Rebased.

rfatykhov-lynx commented:
Hello, @davidtwco! After your approval, we made a few small changes.
Just wanted to check — is everything still okay?

davidtwco commented:
@bors r+

bors commented:
:pushpin: Commit ac4014bd2004811851caea3e81c22cbed7b92aa5 has been approved by `davidtwco`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=davidtwco ac4014bd2004811851caea3e81c22cbed7b92aa5 -->
<!-- homu: {"type":"Approved","sha":"ac4014bd2004811851caea3e81c22cbed7b92aa5","approver":"davidtwco","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit ac4014bd2004811851caea3e81c22cbed7b92aa5 with merge 3920514036f59b5596133707a425f1b8a1c20815...
<!-- homu: {"type":"BuildStarted","head_sha":"ac4014bd2004811851caea3e81c22cbed7b92aa5","merge_sha":"3920514036f59b5596133707a425f1b8a1c20815"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/14503615447/job/40695783632)
Approved by: davidtwco
Pushing 3920514036f59b5596133707a425f1b8a1c20815 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"davidtwco","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/14503615447/job/40695783632"},"merge_sha":"3920514036f59b5596133707a425f1b8a1c20815"} -->

rust-timer commented:
Finished benchmarking commit (3920514036f59b5596133707a425f1b8a1c20815): [comparison URL](https://perf.rust-lang.org/compare.html?start=78f2104e334068d5a892a170d50193c0025c690e&end=3920514036f59b5596133707a425f1b8a1c20815&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=78f2104e334068d5a892a170d50193c0025c690e&end=3920514036f59b5596133707a425f1b8a1c20815&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=78f2104e334068d5a892a170d50193c0025c690e&end=3920514036f59b5596133707a425f1b8a1c20815&stat=max-rss)
<details>
<summary>Results (primary -0.0%, secondary 3.7%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.0%  | [0.7%, 2.6%]   | 3     |
| Regressions ❌ <br /> (secondary)  | 3.7%  | [2.1%, 4.9%]   | 3     |
| Improvements ✅ <br /> (primary)   | -0.7% | [-1.9%, -0.4%] | 8     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.0% | [-1.9%, 2.6%]  | 11    |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=78f2104e334068d5a892a170d50193c0025c690e&end=3920514036f59b5596133707a425f1b8a1c20815&stat=cycles:u)
<details>
<summary>Results (primary 0.4%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.6%  | [0.4%, 0.7%]   | 4     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.5% | [-0.5%, -0.5%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.4%  | [-0.5%, 0.7%]  | 5     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=78f2104e334068d5a892a170d50193c0025c690e&end=3920514036f59b5596133707a425f1b8a1c20815&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 776.148s -> 776.256s (0.01%)
**Artifact size**: 364.77 MiB -> 364.80 MiB (0.01%)
